apex/transformer/pipeline_parallel/p2p_communication.py

```diff
@@ -161,26 +197,59 @@ def _communicate(
             tensor_send_prev = split_tensor_into_1d_equal_chunks(tensor_send_prev)
 
     # Send tensors in both the forward and backward directions as appropriate.
-    _run_p2pops(tensor_send_prev, tensor_send_next, tensor_recv_prev, tensor_recv_next)
-    # To protect against race condition when using batch_isend_irecv().
-    torch.cuda.synchronize()
+    tensor_send_prev_req, tensor_recv_prev_req, tensor_send_next_req, tensor_recv_next_req = _run_p2pops(tensor_send_prev, tensor_send_next, tensor_recv_prev, tensor_recv_next, async_comm=async_comm)
+
+    if async_comm:
+        tensor_recv_prev_waitfunc = None if tensor_recv_prev_req is None else tensor_recv_prev_req.wait
+        tensor_recv_next_waitfunc = None if tensor_recv_next_req is None else tensor_recv_next_req.wait
+        # TODO: investigate whether this is necessary for correctness (ref: https://github.com/pytorch/pytorch/issues/38642)
+        torch.cuda.synchronize()
```

Comment:
Seems that removing this will cause the losses with `async_comm=True` to vary significantly form the `async_comm=False` case.

Reply:
(see updated PR description w/ suspected issue and fix)